File: clients/ios/BuildIt/Modules/CRM/Integrations/CRMCallingIntegration.swift
// CRMCallingIntegration.swift
// BuildIt - Decentralized Mesh Communication
//
// Integration between CRM and Calling modules for caller ID and call history.
// Provides phone lookup, automatic contact creation, and call logging.

import Foundation
import Combine
import os.log

// MARK: - CRM Calling Integration

/// Integration between CRM and Calling modules for caller ID and call history
@MainActor
public class CRMCallingIntegration: ObservableObject {
    // MARK: - Published Properties

    @Published public private(set) var recentCalls: [CallHistoryRecord] = []
    @Published public private(set) var isLoading: Bool = false
    @Published public var lastError: String?

    // MARK: - Private Properties

    // In-memory call history cache (would be persisted in SwiftData in production)
    private var callHistory: [String: [CallHistoryRecord]] = [:] // contactId -> records
    private var contacts: [String: CRMContact] = [:] // id -> contact
    private let logger = Logger(subsystem: "com.buildit", category: "CRMCallingIntegration")

    // MARK: - Initialization

    public init() {}

    // MARK: - Phone Lookup

    /// Look up contact by phone number
    /// - Parameters:
    ///   - phone: Phone number to search
    ///   - groupId: Optional group ID to limit search scope
    /// - Returns: Caller lookup result with contact info if found
    public func lookupByPhone(phone: String, groupId: String? = nil) async throws -> CallerLookupResult {
        let normalizedPhone = normalizePhoneNumber(phone)

        logger.info("Looking up contact by phone: \(normalizedPhone)")

        // Search in local cache first
        for contact in contacts.values {
            if let group = groupId, contact.groupId != group {
                continue
            }

            // Check all phone fields
            if let contactPhone = contact.phone, normalizePhoneNumber(contactPhone) == normalizedPhone {
                let calls = callHistory[contact.id] ?? []
                return CallerLookupResult(
                    found: true,
                    contact: contact,
                    matchedField: .phone,
                    previousCalls: calls.count,
                    lastCallDate: calls.max(by: { $0.startedAt < $1.startedAt })?.startedAt
                )
            }

            if let mobile = contact.mobile, normalizePhoneNumber(mobile) == normalizedPhone {
                let calls = callHistory[contact.id] ?? []
                return CallerLookupResult(
                    found: true,
                    contact: contact,
                    matchedField: .mobile,
                    previousCalls: calls.count,
                    lastCallDate: calls.max(by: { $0.startedAt < $1.startedAt })?.startedAt
                )
            }

            if let workPhone = contact.workPhone, normalizePhoneNumber(workPhone) == normalizedPhone {
                let calls = callHistory[contact.id] ?? []
                return CallerLookupResult(
                    found: true,
                    contact: contact,
                    matchedField: .workPhone,
                    previousCalls: calls.count,
                    lastCallDate: calls.max(by: { $0.startedAt < $1.startedAt })?.startedAt
                )
            }
        }

        // In production, would query SwiftData
        // let descriptor = FetchDescriptor<ContactEntity>(...)
        // let results = try modelContext.fetch(descriptor)

        return CallerLookupResult.notFound
    }

    // MARK: - Contact Creation

    /// Create contact from call
    /// - Parameters:
    ///   - data: Contact data from call
    ///   - groupId: Group to create contact in
    /// - Returns: Created contact
    public func createContactFromCall(
        data: CreateContactFromCallData,
        groupId: String
    ) async throws -> CRMContact {
        let normalizedPhone = normalizePhoneNumber(data.phoneNumber)

        logger.info("Creating contact from call: \(normalizedPhone)")

        let contact = CRMContact(
            id: "contact-\(UUID().uuidString)",
            tableId: "", // Would be set based on CRM template
            groupId: groupId,
            name: data.name ?? "Unknown Caller",
            phone: normalizedPhone,
            customFields: [
                "source": .string("inbound-call"),
                "sourceHotline": data.hotlineId.map { .string($0) } ?? .null,
                "firstContactDate": .int(Int(Date().timeIntervalSince1970)),
                "notes": data.notes.map { .string($0) } ?? .null
            ],
            createdBy: data.operatorPubkey ?? "system",
            updatedBy: data.operatorPubkey ?? "system"
        )

        // Store in cache
        contacts[contact.id] = contact

        // In production, would save to SwiftData
        // let entity = ContactEntity.from(contact)
        // modelContext.insert(entity)
        // try modelContext.save()

        logger.info("Created contact: \(contact.id)")

        return contact
    }

    // MARK: - Call Logging

    /// Log call interaction
    /// - Parameters:
    ///   - contactId: Contact ID to log call for
    ///   - call: Call data (without id/contactId/created)
    /// - Returns: Created call history record
    public func logCallInteraction(
        contactId: String,
        direction: CallDirection,
        phoneNumber: String,
        startedAt: Date,
        endedAt: Date? = nil,
        duration: Int = 0,
        status: CallStatus = .completed,
        recordingUrl: String? = nil,
        transcriptUrl: String? = nil,
        notes: String? = nil,
        operatorPubkey: String? = nil,
        hotlineId: String? = nil
    ) async throws -> CallHistoryRecord {
        let record = CallHistoryRecord(
            id: "call-\(UUID().uuidString)",
            contactId: contactId,
            direction: direction,
            phoneNumber: phoneNumber,
            startedAt: startedAt,
            endedAt: endedAt,
            duration: duration,
            status: status,
            recordingUrl: recordingUrl,
            transcriptUrl: transcriptUrl,
            notes: notes,
            operatorPubkey: operatorPubkey,
            hotlineId: hotlineId,
            created: Date()
        )

        // Store in memory cache
        var contactCalls = callHistory[contactId] ?? []
        contactCalls.append(record)
        callHistory[contactId] = contactCalls

        // Update recent calls
        updateRecentCalls()

        logger.info("Logged call interaction for contact: \(contactId)")
        logger.info("Direction: \(direction.rawValue), Duration: \(duration), Status: \(status.rawValue)")

        // In production, would save to SwiftData
        // let entity = CallHistoryEntity.from(record)
        // modelContext.insert(entity)
        // try modelContext.save()

        return record
    }

    // MARK: - Call History

    /// Get call history for contact
    /// - Parameters:
    ///   - contactId: Contact ID to get history for
    ///   - options: Filter/pagination options
    /// - Returns: List of call history records
    public func getContactCallHistory(
        contactId: String,
        options: CallHistoryOptions? = nil
    ) async throws -> [CallHistoryRecord] {
        var calls = callHistory[contactId] ?? []

        // Apply filters
        if let opts = options {
            if let direction = opts.direction {
                calls = calls.filter { $0.direction == direction }
            }

            if let dateFrom = opts.dateFrom {
                calls = calls.filter { $0.startedAt >= dateFrom }
            }

            if let dateTo = opts.dateTo {
                calls = calls.filter { $0.startedAt <= dateTo }
            }
        }

        // Sort by date descending (most recent first)
        calls.sort { $0.startedAt > $1.startedAt }

        // Apply pagination
        if let opts = options {
            if let offset = opts.offset {
                calls = Array(calls.dropFirst(offset))
            }

            if let limit = opts.limit {
                calls = Array(calls.prefix(limit))
            }
        }

        return calls
    }

    // MARK: - Engagement Score

    /// Update engagement score after call
    /// - Parameters:
    ///   - contactId: Contact ID
    ///   - callDuration: Duration of call in seconds
    ///   - direction: Call direction
    /// - Returns: Engagement update result
    public func updateEngagementFromCall(
        contactId: String,
        callDuration: Int,
        direction: CallDirection
    ) async throws -> CallEngagementUpdate {
        // Get current engagement score (would fetch from database)
        let previousScore = contacts[contactId]?.customFields["engagement_score"]?.intValue ?? 0

        // Calculate score increase based on call metrics
        var scoreIncrease = 0

        // Base points for having a call
        scoreIncrease += direction == .inbound ? 10 : 5

        // Additional points based on duration (capped)
        let durationMinutes = callDuration / 60
        scoreIncrease += min(durationMinutes * 2, 20)

        let newScore = previousScore + scoreIncrease

        logger.info("Updated engagement score for contact: \(contactId)")
        logger.info("Previous: \(previousScore), New: \(newScore), Change: \(scoreIncrease)")

        // In production, would update in database
        if var contact = contacts[contactId] {
            contact.customFields["engagement_score"] = .int(newScore)
            contacts[contactId] = contact
        }

        return CallEngagementUpdate(
            contactId: contactId,
            previousScore: previousScore,
            newScore: newScore,
            callDuration: callDuration,
            callDirection: direction
        )
    }

    // MARK: - Call Statistics

    /// Get call statistics for a contact
    /// - Parameter contactId: Contact ID
    /// - Returns: Call statistics
    public func getContactCallStats(contactId: String) async throws -> ContactCallStats {
        let calls = callHistory[contactId] ?? []

        guard !calls.isEmpty else {
            return ContactCallStats.empty
        }

        let inboundCalls = calls.filter { $0.direction == .inbound }.count
        let outboundCalls = calls.filter { $0.direction == .outbound }.count
        let totalDuration = calls.reduce(0) { $0 + $1.duration }
        let lastCallDate = calls.max(by: { $0.startedAt < $1.startedAt })?.startedAt
        let missedCalls = calls.filter { $0.status == .missed }.count

        return ContactCallStats(
            totalCalls: calls.count,
            inboundCalls: inboundCalls,
            outboundCalls: outboundCalls,
            totalDuration: totalDuration,
            averageDuration: totalDuration / calls.count,
            lastCallDate: lastCallDate,
            missedCalls: missedCalls
        )
    }

    // MARK: - Recent Calls

    /// Get recent calls across all contacts
    /// - Parameters:
    ///   - groupId: Optional group filter
    ///   - limit: Maximum number of calls to return
    /// - Returns: Recent call records
    public func getRecentCalls(
        groupId: String? = nil,
        limit: Int = 50
    ) async throws -> [CallHistoryRecord] {
        var allCalls: [CallHistoryRecord] = []

        for calls in callHistory.values {
            allCalls.append(contentsOf: calls)
        }

        // Filter by group if specified
        if let groupId = groupId {
            let groupContactIds = Set(contacts.values.filter { $0.groupId == groupId }.map { $0.id })
            allCalls = allCalls.filter { groupContactIds.contains($0.contactId) }
        }

        // Sort by date descending
        allCalls.sort { $0.startedAt > $1.startedAt }

        return Array(allCalls.prefix(limit))
    }

    // MARK: - Recording & Notes

    /// Link call recording to contact
    /// - Parameters:
    ///   - contactId: Contact ID
    ///   - callId: Call ID
    ///   - recordingUrl: URL of the recording
    ///   - transcriptUrl: Optional URL of the transcript
    public func linkCallRecording(
        contactId: String,
        callId: String,
        recordingUrl: String,
        transcriptUrl: String? = nil
    ) async throws {
        guard var contactCalls = callHistory[contactId],
              let index = contactCalls.firstIndex(where: { $0.id == callId }) else {
            logger.warning("Call not found: \(callId) for contact: \(contactId)")
            return
        }

        var record = contactCalls[index]
        record.recordingUrl = recordingUrl
        if let transcript = transcriptUrl {
            record.transcriptUrl = transcript
        }
        contactCalls[index] = record
        callHistory[contactId] = contactCalls

        logger.info("Linked recording to call: \(callId)")
    }

    /// Add notes to a call
    /// - Parameters:
    ///   - contactId: Contact ID
    ///   - callId: Call ID
    ///   - notes: Notes to add
    public func addCallNotes(
        contactId: String,
        callId: String,
        notes: String
    ) async throws {
        guard var contactCalls = callHistory[contactId],
              let index = contactCalls.firstIndex(where: { $0.id == callId }) else {
            logger.warning("Call not found: \(callId) for contact: \(contactId)")
            return
        }

        var record = contactCalls[index]
        record.notes = notes
        contactCalls[index] = record
        callHistory[contactId] = contactCalls

        logger.info("Added notes to call: \(callId)")
    }

    // MARK: - Contact Management

    /// Add contact to cache (for testing/initialization)
    public func addContact(_ contact: CRMContact) {
        contacts[contact.id] = contact
    }

    /// Get contact by ID
    public func getContact(id: String) -> CRMContact? {
        contacts[id]
    }

    // MARK: - Module Status

    /// Check if calling module is available
    public func isCallingModuleAvailable() async -> Bool {
        // In production, would check if calling module is enabled and configured
        return true
    }

    // MARK: - Private Helpers

    private func updateRecentCalls() {
        var allCalls: [CallHistoryRecord] = []
        for calls in callHistory.values {
            allCalls.append(contentsOf: calls)
        }
        allCalls.sort { $0.startedAt > $1.startedAt }
        recentCalls = Array(allCalls.prefix(50))
    }

    /// Normalize phone number for consistent lookup
    private func normalizePhoneNumber(_ phone: String) -> String {
        // Remove all non-digit characters except leading +
        let hasPlus = phone.hasPrefix("+")
        let digits = phone.filter { $0.isNumber }

        // For US numbers, ensure E.164 format
        if digits.count == 10 {
            return "+1\(digits)"
        }

        if digits.count == 11 && digits.hasPrefix("1") {
            return "+\(digits)"
        }

        // Return with + prefix if originally had it
        return hasPlus ? "+\(digits)" : digits
    }
}

// MARK: - Errors

/// Errors for CRM calling integration
public enum CRMCallingError: LocalizedError, Sendable {
    case contactNotFound
    case callNotFound
    case invalidPhoneNumber
    case callingModuleUnavailable

    public var errorDescription: String? {
        switch self {
        case .contactNotFound:
            return "Contact not found"
        case .callNotFound:
            return "Call record not found"
        case .invalidPhoneNumber:
            return "Invalid phone number format"
        case .callingModuleUnavailable:
            return "Calling module is not available"
        }
    }
}

File: clients/android/app/src/main/java/network/buildit/modules/crm/integration/CRMCallingIntegration.kt
package network.buildit.modules.crm.integration

import android.util.Log
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.Serializable
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Integration between CRM and Calling modules for caller ID and call history.
 *
 * Provides functionality for:
 * - Looking up contacts by phone number
 * - Creating contacts from calls
 * - Logging call interactions
 * - Tracking call history per contact
 * - Updating engagement scores based on calls
 * - Linking call recordings to contacts
 */
@Singleton
class CRMCallingIntegration @Inject constructor() {

    companion object {
        private const val TAG = "CRMCallingIntegration"

        // Engagement score adjustments
        private const val INBOUND_CALL_POINTS = 5
        private const val OUTBOUND_CALL_POINTS = 3
        private const val CALL_DURATION_POINTS_PER_MINUTE = 1
        private const val MAX_CALL_DURATION_POINTS = 10
    }

    // In-memory storage for call history (in production, this would use Room database)
    private val callHistoryByContact = ConcurrentHashMap<String, MutableList<CallHistoryRecord>>()

    // Contact phone number index for lookups
    private val phoneToContactIndex = ConcurrentHashMap<String, ContactPhoneEntry>()

    // Engagement scores
    private val engagementScores = ConcurrentHashMap<String, Int>()

    // State flows for observation
    private val _recentCallActivity = MutableStateFlow<List<CallHistoryRecord>>(emptyList())
    val recentCallActivity: StateFlow<List<CallHistoryRecord>> = _recentCallActivity.asStateFlow()

    /**
     * Look up contact by phone number.
     *
     * @param phone Phone number to search for.
     * @param groupId Optional group ID to filter by.
     * @return Caller lookup result with contact info if found.
     */
    suspend fun lookupByPhone(phone: String, groupId: String? = null): CallerLookupResult {
        val normalizedPhone = normalizePhoneNumber(phone)

        Log.d(TAG, "Looking up contact by phone: $normalizedPhone")

        val entry = phoneToContactIndex[normalizedPhone]
        if (entry == null) {
            return CallerLookupResult(
                found = false,
                contact = null,
                matchedField = null,
                previousCalls = null,
                lastCallDate = null
            )
        }

        // Filter by group if specified
        if (groupId != null && entry.contact.groupId != groupId) {
            return CallerLookupResult(
                found = false,
                contact = null,
                matchedField = null,
                previousCalls = null,
                lastCallDate = null
            )
        }

        // Get call history for this contact
        val callHistory = callHistoryByContact[entry.contact.id] ?: emptyList()
        val lastCall = callHistory.maxByOrNull { it.startedAt }

        return CallerLookupResult(
            found = true,
            contact = entry.contact,
            matchedField = entry.fieldType,
            previousCalls = callHistory.size,
            lastCallDate = lastCall?.startedAt
        )
    }

    /**
     * Create contact from call information.
     *
     * @param data Data from the call.
     * @param groupId Group to add the contact to.
     * @return Created CRM contact.
     */
    suspend fun createContactFromCall(
        data: CreateContactFromCallData,
        groupId: String
    ): CRMContact {
        val contactId = UUID.randomUUID().toString()
        val now = System.currentTimeMillis()

        val contact = CRMContact(
            id = contactId,
            groupId = groupId,
            name = data.name,
            phone = data.phoneNumber,
            mobile = null,
            workPhone = null,
            email = null,
            notes = data.notes,
            engagementScore = 0,
            createdAt = now,
            updatedAt = now,
            source = ContactSource.CALL,
            sourceDetails = buildString {
                append("Created from call")
                data.hotlineId?.let { append(" via hotline $it") }
                data.operatorPubkey?.let { append(" by operator $it") }
            }
        )

        // Index the phone number
        val normalizedPhone = normalizePhoneNumber(data.phoneNumber)
        phoneToContactIndex[normalizedPhone] = ContactPhoneEntry(
            contact = contact,
            fieldType = PhoneFieldType.PHONE
        )

        Log.i(TAG, "Created contact $contactId from call: ${data.phoneNumber}")
        return contact
    }

    /**
     * Register existing contact's phone numbers for lookup.
     *
     * @param contact The contact to register.
     */
    fun registerContactPhones(contact: CRMContact) {
        contact.phone?.let { phone ->
            phoneToContactIndex[normalizePhoneNumber(phone)] = ContactPhoneEntry(contact, PhoneFieldType.PHONE)
        }
        contact.mobile?.let { mobile ->
            phoneToContactIndex[normalizePhoneNumber(mobile)] = ContactPhoneEntry(contact, PhoneFieldType.MOBILE)
        }
        contact.workPhone?.let { workPhone ->
            phoneToContactIndex[normalizePhoneNumber(workPhone)] = ContactPhoneEntry(contact, PhoneFieldType.WORK_PHONE)
        }
    }

    /**
     * Log call interaction to contact's history.
     *
     * @param contactId The contact ID.
     * @param call Call history record.
     * @return The saved call history record.
     */
    suspend fun logCallInteraction(
        contactId: String,
        call: CallHistoryRecord
    ): CallHistoryRecord {
        val recordWithId = if (call.id.isEmpty()) {
            call.copy(id = UUID.randomUUID().toString())
        } else {
            call
        }

        val history = callHistoryByContact.getOrPut(contactId) { mutableListOf() }
        history.add(recordWithId)

        // Update recent activity
        updateRecentActivity()

        Log.i(TAG, "Logged call interaction for contact $contactId: ${recordWithId.id}")
        return recordWithId
    }

    /**
     * Get call history for a contact.
     *
     * @param contactId The contact ID.
     * @param options Optional filtering options.
     * @return Flow of call history records.
     */
    fun getContactCallHistory(
        contactId: String,
        options: CallHistoryOptions? = null
    ): Flow<List<CallHistoryRecord>> {
        return flow {
            var records = callHistoryByContact[contactId]?.toList() ?: emptyList()

            // Apply filters
            options?.let { opts ->
                opts.direction?.let { dir ->
                    records = records.filter { it.direction == dir }
                }
                opts.dateFrom?.let { from ->
                    records = records.filter { it.startedAt >= from }
                }
                opts.dateTo?.let { to ->
                    records = records.filter { it.startedAt <= to }
                }
            }

            // Sort by date descending
            records = records.sortedByDescending { it.startedAt }

            // Apply pagination
            options?.let { opts ->
                opts.offset?.let { offset ->
                    records = records.drop(offset)
                }
                opts.limit?.let { limit ->
                    records = records.take(limit)
                }
            }

            emit(records)
        }
    }

    /**
     * Update engagement score after a call.
     *
     * @param contactId The contact ID.
     * @param callDuration Call duration in seconds.
     * @param direction Call direction.
     * @return Engagement update details.
     */
    suspend fun updateEngagementFromCall(
        contactId: String,
        callDuration: Int,
        direction: CallDirection
    ): CallEngagementUpdate {
        val previousScore = engagementScores.getOrDefault(contactId, 0)

        // Calculate points
        var points = when (direction) {
            CallDirection.INBOUND -> INBOUND_CALL_POINTS
            CallDirection.OUTBOUND -> OUTBOUND_CALL_POINTS
        }

        // Add duration bonus (capped)
        val durationMinutes = callDuration / 60
        val durationPoints = minOf(durationMinutes * CALL_DURATION_POINTS_PER_MINUTE, MAX_CALL_DURATION_POINTS)
        points += durationPoints

        val newScore = previousScore + points
        engagementScores[contactId] = newScore

        Log.d(TAG, "Updated engagement for contact $contactId: $previousScore -> $newScore (+$points)")

        return CallEngagementUpdate(
            contactId = contactId,
            previousScore = previousScore,
            newScore = newScore,
            callDuration = callDuration,
            callDirection = direction
        )
    }

    /**
     * Get call statistics for a contact.
     *
     * @param contactId The contact ID.
     * @return Aggregated call statistics.
     */
    suspend fun getContactCallStats(contactId: String): ContactCallStats {
        val records = callHistoryByContact[contactId] ?: emptyList()

        if (records.isEmpty()) {
            return ContactCallStats(
                contactId = contactId,
                totalCalls = 0,
                inboundCalls = 0,
                outboundCalls = 0,
                totalDuration = 0,
                averageDuration = 0,
                lastCallDate = null,
                missedCalls = 0
            )
        }

        val inbound = records.filter { it.direction == CallDirection.INBOUND }
        val outbound = records.filter { it.direction == CallDirection.OUTBOUND }
        val missed = records.filter { it.status == CallStatus.MISSED }
        val completed = records.filter { it.status == CallStatus.COMPLETED }

        val totalDuration = completed.sumOf { it.duration }
        val avgDuration = if (completed.isNotEmpty()) totalDuration / completed.size else 0

        return ContactCallStats(
            contactId = contactId,
            totalCalls = records.size,
            inboundCalls = inbound.size,
            outboundCalls = outbound.size,
            totalDuration = totalDuration,
            averageDuration = avgDuration,
            lastCallDate = records.maxOfOrNull { it.startedAt },
            missedCalls = missed.size
        )
    }

    /**
     * Link call recording to a contact's call history.
     *
     * @param contactId The contact ID.
     * @param callId The call ID.
     * @param recordingUrl URL to the recording.
     * @param transcriptUrl Optional URL to the transcript.
     */
    suspend fun linkCallRecording(
        contactId: String,
        callId: String,
        recordingUrl: String,
        transcriptUrl: String? = null
    ) {
        val history = callHistoryByContact[contactId] ?: return
        val index = history.indexOfFirst { it.id == callId }

        if (index >= 0) {
            val updated = history[index].copy(
                recordingUrl = recordingUrl,
                transcriptUrl = transcriptUrl
            )
            history[index] = updated
            Log.i(TAG, "Linked recording to call $callId for contact $contactId")
        } else {
            Log.w(TAG, "Call $callId not found for contact $contactId")
        }
    }

    /**
     * Add notes to a call.
     *
     * @param contactId The contact ID.
     * @param callId The call ID.
     * @param notes Notes to add.
     */
    suspend fun addCallNotes(
        contactId: String,
        callId: String,
        notes: String
    ) {
        val history = callHistoryByContact[contactId] ?: return
        val index = history.indexOfFirst { it.id == callId }

        if (index >= 0) {
            val existing = history[index].notes
            val updated = history[index].copy(
                notes = if (existing.isNullOrEmpty()) notes else "$existing\n$notes"
            )
            history[index] = updated
            Log.i(TAG, "Added notes to call $callId for contact $contactId")
        }
    }

    /**
     * Get engagement score for a contact.
     *
     * @param contactId The contact ID.
     * @return Current engagement score.
     */
    fun getEngagementScore(contactId: String): Int {
        return engagementScores.getOrDefault(contactId, 0)
    }

    /**
     * Observe call history for a contact.
     *
     * @param contactId The contact ID.
     * @return Flow of call history records.
     */
    fun observeContactCallHistory(contactId: String): Flow<List<CallHistoryRecord>> {
        return flow {
            val records = callHistoryByContact[contactId]?.sortedByDescending { it.startedAt }
                ?: emptyList()
            emit(records)
        }
    }

    private fun normalizePhoneNumber(phone: String): String {
        // Remove all non-digit characters except leading +
        val hasPlus = phone.startsWith("+")
        val digitsOnly = phone.filter { it.isDigit() }
        return if (hasPlus) "+$digitsOnly" else digitsOnly
    }

    private fun updateRecentActivity() {
        val allRecords = callHistoryByContact.values.flatten()
            .sortedByDescending { it.startedAt }
            .take(50)
        _recentCallActivity.value = allRecords
    }
}

/**
 * Result of looking up a contact by phone.
 */
data class CallerLookupResult(
    val found: Boolean,
    val contact: CRMContact?,
    val matchedField: PhoneFieldType?,
    val previousCalls: Int?,
    val lastCallDate: Long?
)

/**
 * Phone field types for matching.
 */
enum class PhoneFieldType {
    PHONE,
    MOBILE,
    WORK_PHONE
}

/**
 * CRM Contact representation for calling integration.
 */
@Serializable
data class CRMContact(
    val id: String,
    val groupId: String,
    val name: String?,
    val phone: String?,
    val mobile: String?,
    val workPhone: String?,
    val email: String?,
    val notes: String?,
    val engagementScore: Int,
    val createdAt: Long,
    val updatedAt: Long,
    val source: ContactSource? = null,
    val sourceDetails: String? = null
)

/**
 * Contact source.
 */
enum class ContactSource {
    CALL,
    IMPORT,
    FORM,
    MANUAL
}

/**
 * Internal index entry for phone lookup.
 */
internal data class ContactPhoneEntry(
    val contact: CRMContact,
    val fieldType: PhoneFieldType
)

/**
 * Call history record.
 */
@Serializable
data class CallHistoryRecord(
    val id: String,
    val contactId: String,
    val direction: CallDirection,
    val phoneNumber: String,
    val startedAt: Long,
    val endedAt: Long?,
    val duration: Int,
    val status: CallStatus,
    val recordingUrl: String?,
    val transcriptUrl: String?,
    val notes: String?,
    val operatorPubkey: String?,
    val hotlineId: String?,
    val created: Long
)

/**
 * Call direction.
 */
enum class CallDirection {
    INBOUND,
    OUTBOUND
}

/**
 * Call status.
 */
enum class CallStatus {
    COMPLETED,
    MISSED,
    VOICEMAIL,
    FAILED
}

/**
 * Options for filtering call history.
 */
data class CallHistoryOptions(
    val limit: Int? = null,
    val offset: Int? = null,
    val direction: CallDirection? = null,
    val dateFrom: Long? = null,
    val dateTo: Long? = null
)

/**
 * Result of engagement score update.
 */
data class CallEngagementUpdate(
    val contactId: String,
    val previousScore: Int,
    val newScore: Int,
    val callDuration: Int,
    val callDirection: CallDirection
)

/**
 * Aggregated call statistics for a contact.
 */
data class ContactCallStats(
    val contactId: String,
    val totalCalls: Int,
    val inboundCalls: Int,
    val outboundCalls: Int,
    val totalDuration: Int,
    val averageDuration: Int,
    val lastCallDate: Long?,
    val missedCalls: Int
)

/**
 * Data for creating a contact from a call.
 */
data class CreateContactFromCallData(
    val phoneNumber: String,
    val name: String? = null,
    val notes: String? = null,
    val hotlineId: String? = null,
    val operatorPubkey: String? = null
)
